certora/specs/RatioMath.spec

```diff
@@ -0,0 +1,157 @@
+    function lastUpdate(MorphoHarness.Id) external returns uint256 envfree;
+
+    function maxFee() external returns uint256 envfree;
+    function libId(MorphoHarness.MarketParams) external returns MorphoHarness.Id envfree;
+
+    function MathLib.mulDivDown(uint256 a, uint256 b, uint256 c) internal returns uint256 => summaryMulDivDown(a,b,c);
+    function MathLib.mulDivUp(uint256 a, uint256 b, uint256 c) internal returns uint256 => summaryMulDivUp(a,b,c);
+    function MathLib.wTaylorCompounded(uint256, uint256) internal returns uint256 => NONDET;
+
+    function _.borrowRate(MorphoHarness.MarketParams, MorphoHarness.Market) external => HAVOC_ECF;
+
+}
+
+invariant feeInRange(MorphoHarness.Id id)
+    fee(id) <= maxFee();
+
+// This is a simple overapproximative summary, stating that it rounds in the right direction.
+function summaryMulDivUp(uint256 x, uint256 y, uint256 d) returns uint256 {
+    uint256 result;
+    // Safe require that is checked by the specification in LibSummary.spec.
+    require result * d >= x * y;
+    return result;
+}
+
+// This is a simple overapproximative summary, stating that it rounds in the right direction.
+function summaryMulDivDown(uint256 x, uint256 y, uint256 d) returns uint256 {
+    uint256 result;
+    // Safe require that is checked by the specification in LibSummary.spec.
+    require result * d <= x * y;
+    return result;
+}
+
+// Check that accrueInterest increases the value of supply shares.
+rule accrueInterestIncreasesSupplyRatio(env e, MorphoHarness.MarketParams marketParams) {
+    MorphoHarness.Id id;
+    requireInvariant feeInRange(id);
+
+    mathint assetsBefore = virtualTotalSupplyAssets(id);
+    mathint sharesBefore = virtualTotalSupplyShares(id);
+
+    // The check is done for every market, not just for id.
+    accrueInterest(e, marketParams);
+
+    mathint assetsAfter = virtualTotalSupplyAssets(id);
+    mathint sharesAfter = virtualTotalSupplyShares(id);
+
+    // Check that the ratio increases: assetsBefore/sharesBefore <= assetsAfter / sharesAfter.
+    assert assetsBefore * sharesAfter <= assetsAfter * sharesBefore;
+}
+
+// Check that accrueInterest increases the value of borrow shares.
+rule accrueInterestIncreasesBorrowRatio(env e, MorphoHarness.MarketParams marketParams) {
+    MorphoHarness.Id id;
+    requireInvariant feeInRange(id);
+
+    mathint assetsBefore = virtualTotalBorrowAssets(id);
+    mathint sharesBefore = virtualTotalBorrowShares(id);
+
+    // The check is done for every marketParams, not just for id.
+    accrueInterest(e, marketParams);
+
+    mathint assetsAfter = virtualTotalBorrowAssets(id);
+    mathint sharesAfter = virtualTotalBorrowShares(id);
+
+    // Check that the ratio increases: assetsBefore/sharesBefore <= assetsAfter / sharesAfter.
+    assert assetsBefore * sharesAfter <= assetsAfter * sharesBefore;
+}
+
+
+// Check that excepti for liquidate, every function increases the value of supply shares.
```

Comment:
Yes it's not intuitive but it's a nice property: other user's cannot decrease the value of your supply shares. This is because we round in favor of the protocol (so in favor of passive users actually).

This was already challenging to verify, and required to use the `summaryMulDivUp` and `summaryMulDivDown` which are now kind of tailored for this rule, so it most likely wouldn't work when bounding the increase